File: assignment-1/codes/problem_4a.py
# CS21BTECH11013
# Hema Sri Cheekatla
# March 29, 2022
# 2018 Paper, Problem 4a
import matplotlib.pyplot as plt
import numpy as np

# 	-2+10x <= 13x+10 < 10x+24
# collecting the set of integers obeying the given expression
set = list()
n = -1000
k = 0
#checking the numbers from -1000 to 1000
# we can take any appropriate range
while n < 1001:
	a = -2 + 10*n
	b = 13*n + 10
	c = 24 + 10*n
	if a<=b and b<c:
		k = k + 1
		set.append(n)
	n = n + 1

# y cordinates for plotting points on real number line graph
# y cordinates are obviously 0s since we are plotting integers on real axis
y = list()
while k > 0:
	y.append(0)
	k = k - 1


x = np.linspace(-6, 6, 1000)
y_1 = -2 + 10*x
l1 = [10, -2]
plt.plot(x, y_1, 'b', label = "y1")
y_2 = 13*x + 10
l2 = [13, 10]
plt.plot(x, y_2, 'g', label = "y2")
y_3 = 24 + 10*x
l3 = [10, 24]
plt.plot(x, y_3, 'r', label = "y3")
plt.xlabel('x-axis')
plt.ylabel('y-axis')


""" the values of x that obey the inequality y1 <= y2 < y3 are
 the values at where the line y2 lies between y1 and y3
 since y2 intesects y1 at x = -4 and y3 at x = 14/3,
 the possible values of x are found to be the integers in the set [-4, 4.67]
 if y = mx + c consider the list [m,c] to find the intersection points
"""
def intersec(a,b,c,d):
	xi = (b-d)/(c-a)
	yi = a*xi + b
	return xi, yi
# (x1,y1) intersection point of y1 and y2
# (x2,y2) intersection point of y2 and y3
x1, y1 = intersec(l1[0], l1[1],l2[0], l2[1])
x2, y2 = intersec(l2[0], l2[1], l3[0], l3[1])
plt.fill_between(x,y_1,y_3, where=[(x>=x1) and (x<x2) for x in x], color = 'green', alpha=0.3, label = 'region at where y1 <= y2 < y3')

plt.scatter(x1, y1, c = 'black')
plt.axvline(x = x1, color = 'grey', linestyle = '--')

plt.scatter(x2, y2, c = 'black')
plt.axvline(x = x2, color = 'grey', linestyle = '--')
plt.axhline(y=0, color = 'black')
plt.axvline(x=0, color = 'black')
plt.scatter(set, y, c = "blue", label = "x co-ordinates in this region")

plt.grid()
plt.legend(loc = "upper left")
plt.title('Graph of corresponding lines')
plt.savefig("../images/Figure_1.png")
plt.show()

plt.scatter(set, y, c = "blue", label = "points in the given set")

plt.xlabel('real - axis')

# limitting x axis since we are getting integral solutions between -4 and 4
plt.xlim(-6, 6)
# we can neglect y axis since we just need points on x axis
plt.ylim(0,0.5)
plt.legend(loc = "upper left")
plt.title('Pointing graph')
plt.savefig("../images/Figure_2.png")
plt.show()
print('Done!')

File: assignment-1/codes/Problem_4a.c
/* Name : Hema Sri Cheekatla
CS21BTECH11013
2018 Paper, problem 4a
Solve the system of linear equations to find the set of integral values of x,
for which the following inequation holds,

	10x-2 <= 13x+10 < 10x+24, where x belongs to integers
line 1 = 10x-y = 2
line 2 = 13x-y = -10
line 3 = 10x-y = -24 */

# include <stdio.h>
// a struct to store two values (touple, coordinates/coefficients )
struct coord{
	float x;
	float y;
};

// making touple
struct coord make_coord(float a, float b){
	struct coord n;
	n.x = a;
	n.y = b;
	return n;
}

// solving equations by using the coefficient matrix and the constants r and s
struct coord solve(struct coord m, struct coord n, float r, float s){
	struct coord p;
	p.x = (r*(n.y) - s*(m.y))/((m.x)*(n.y) - (n.x)*(m.y));
	p.y = (r*(n.x) - s*(m.x))/((n.x)*(m.y) - (m.x)*(n.y));
	return p;
}
// Printing the pair
void print_coord(struct coord p){
	printf("( %.2f, %.2f)\n", p.x, p.y);
}
int main(){
	struct coord a1, a2, a3;
	float b1, b2, b3;
	a1 = make_coord(10, -1); // coefficients of line 1
	a2 = make_coord(13, -1); // coefficients of line 2
	a3 = make_coord(10, -1); // coefficients of line 3
	b1 = 2;
	b2 = -10;
	b3 = -24;
// p1 and p2 intersection points
	struct coord p1, p2;
// solving line 1 and line 2
	p1 = solve(a1,a2,b1,b2);
// solving line 2 and line 3
	p2 = solve(a2,a3,b2,b3);
	printf("The point of intersection of line 1 and line 2 :");
	print_coord(p1);
	printf("The point of intersection of line 3 and line 2 :");
	print_coord(p2);
// set of required integral x values
	printf("The set of x coordinates satisfying the required condition are: \n");
	int i;
	for (i=(int)p1.x; i<p2.x; i++) printf("%d, ",i);
	printf("\n");
	return 0;
}
